### PR TITLE
BFloat16: add explicit dtype support for to_mkldnn and to_dense

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48994 enable mkldnn conv2d backward to support mkldnn tensor input
* #48922 BFloat16: enable prepacked weights's inference
* **#48881 BFloat16: add explicit dtype support for to_mkldnn and to_dense**

Differential Revision: [D25537190](https://our.internmc.facebook.com/intern/diff/D25537190)

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
Rebased version of https://github.com/pytorch/pytorch/pull/37215 .

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 1]
```diff
@@ -343,9 +343,10 @@ SparseTensor dense_to_sparse(const Tensor& self, int64_t sparse_dim){
 
 // NB: Dropped the resizeNd variants
 
-Tensor sparse_to_dense(const SparseTensor& self) {
+Tensor sparse_to_dense(const SparseTensor& self, c10::optional<ScalarType> dtype) {
+  TORCH_CHECK(!dtype.has_value(), "dtype is not need provided for sparse_to_dense");
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:347]: super nit: dtype argument is not supported by sparse_to_dense

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by ngimel** [comment]
xla error looks related (it's bfloat16)

**Comment by VitalyFedyunin** [comment]
Please rebase as it will likely fix xla tests plus internal tools require more recent code.

**Pull request closed without merging**